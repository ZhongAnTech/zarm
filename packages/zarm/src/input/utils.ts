export function getValue(props, defaultValue: string | number = ''): string | number {
  if (typeof props.value !== 'undefined' && props.value !== null) {
    return props.value;
  }
  if (typeof props.defaultValue !== 'undefined' && props.defaultValue !== null) {
    return props.defaultValue;
  }
  return defaultValue;
}
