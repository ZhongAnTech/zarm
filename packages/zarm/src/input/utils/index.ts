export const regexAstralSymbols = /[\uD800-\uDBFF][\uDC00-\uDFFF]|\n/g;

export const countSymbols = (text = '') => {
  return text.replace(regexAstralSymbols, '_').length;
};
