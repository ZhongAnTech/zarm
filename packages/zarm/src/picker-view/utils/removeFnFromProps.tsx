import omit from 'lodash/omit';

const removeFnFromProps = (props, fnList) => {
  return omit(props, fnList);
};

export default removeFnFromProps;
