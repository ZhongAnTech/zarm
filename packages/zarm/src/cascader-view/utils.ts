import { FieldNames } from '../picker-view/interface';
import { resolvedFieldNames } from '../picker-view/utils';
import type {
  BaseCascaderViewProps,
  CascaderItem,
  CascaderOption,
  CascaderValue,
} from './interface';

export const parseState = (props: BaseCascaderViewProps) => {
  const { value, defaultValue } = props;
  const currentValue: CascaderValue[] = [];

  return {
    value: value || defaultValue,
    currentValue,
  };
};

export const parseItems = (
  options: CascaderOption[],
  value: CascaderValue[],
  fieldNames?: Partial<FieldNames>,
) => {
  const field = resolvedFieldNames(fieldNames);
  const items: CascaderItem[] = [];
  const traverse = (memo: CascaderOption[], depth = 0) => {
    memo.forEach((option) => {
      const children = option[field.children] as CascaderOption[] | undefined;
      const currentValue = value[depth];
      if (option[field.value] === currentValue) {
        items[depth] = option;

        if (Array.isArray(children) && children.length > 0) {
          traverse(children, depth + 1);
        }
      }
    });
  };

  traverse(options);

  return items;
};
