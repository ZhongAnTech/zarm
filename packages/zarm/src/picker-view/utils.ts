import omit from 'lodash/omit';
import { toArray } from '../utils';
import type {
  BasePickerViewProps as PickerViewProps,
  PickerViewColumn,
  PickerViewColumnItem,
  PickerViewDataSource,
  PickerViewOption,
  PickerViewValue,
} from './interface';

const DEFAULT_FIELD_NAMES = {
  value: 'value',
  label: 'label',
  children: 'children',
};

export const resolvedFieldNames = <T = object>(
  left: Partial<T> | undefined,
  right: Required<T>,
) => {
  const merged = { ...right };
  Object.keys(right).forEach((key) => {
    merged[key] = left?.[key] || right[key];
  });
  return merged;
};

export const isCascader = (dataSource?: PickerViewDataSource): dataSource is PickerViewOption[] => {
  return Array.isArray(dataSource) && dataSource[0] && !Array.isArray(dataSource[0]);
};
export const isColumn = (dataSource?: PickerViewDataSource): dataSource is PickerViewColumn => {
  return Array.isArray(dataSource) && dataSource[0] && Array.isArray(dataSource[0]);
};

export const isValidValue = (value?: PickerViewValue | PickerViewValue[]) => {
  const currentValue = toArray(value);
  return currentValue.some((item) => !!item || item === 0 || item === false);
};

const resolvedValue = (props: PickerViewProps, initialValue?: PickerViewValue[]) => {
  const { value, defaultValue, wheelDefaultValue } = props;
  if ('value' in props && isValidValue(value)) {
    return toArray(value);
  }

  if ('defaultValue' in props && isValidValue(defaultValue)) {
    return toArray(defaultValue);
  }

  if ('wheelDefaultValue' in props && isValidValue(wheelDefaultValue)) {
    return toArray(wheelDefaultValue);
  }

  return toArray(initialValue);
};

export const resolveColumn = (props: PickerViewProps) => {
  const columns = toArray(props.dataSource) as PickerViewColumn;
  const fieldNames = resolvedFieldNames(props.fieldNames, DEFAULT_FIELD_NAMES);
  const value = resolvedValue(
    props,
    columns.map((item) => item?.[0]?.[fieldNames?.value!]),
  );
  return {
    value,
    columns,
    items: columns.map(
      (column, index) => column.filter((option) => option?.[fieldNames.value] === value[index])[0],
    ),
  };
};

const resolveCascade = (props: PickerViewProps) => {
  const { cols } = props;
  const fieldNames = resolvedFieldNames(props.fieldNames, DEFAULT_FIELD_NAMES);

  const value: PickerViewValue[] = resolvedValue(props, []);
  const columns: PickerViewColumn[] = [];
  const items: PickerViewColumnItem[] = [];
  const traverse = (options: PickerViewOption[], depth = 0) => {
    columns[depth] = options.map((option, index) => {
      const rest = omit<PickerViewColumnItem>(option, fieldNames.children) as PickerViewColumnItem;
      const children = option[fieldNames.children] as PickerViewOption[] | undefined;
      const currentValue = value[depth];
      if (
        (isValidValue(currentValue) && rest[fieldNames.value] === currentValue) ||
        (!isValidValue(currentValue) && index === 0)
      ) {
        value[depth] = rest[fieldNames.value];
        items[depth] = rest;

        if (Array.isArray(children) && children.length > 0 && depth + 1 < cols!) {
          traverse(children, depth + 1);
        }
      }
      return rest;
    });
  };

  traverse(props.dataSource as PickerViewOption[]);

  return {
    value,
    items,
    columns,
  };
};

export const resolved = (props: PickerViewProps) => {
  return isCascader(props.dataSource) ? resolveCascade(props) : resolveColumn(props);
};
