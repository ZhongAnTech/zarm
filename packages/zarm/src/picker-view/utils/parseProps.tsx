import { isArray, isCascader } from '../../utils/validate';

export const isValueValid = (value) => {
  const valueArray = [].concat(value);
  return valueArray.some((item) => !!item || item === 0 || item === false);
};

const getValues = (props, defaultValue?: any) => {
  if ('value' in props && isValueValid(props.value)) {
    return [].concat(props.value);
  }

  if ('defaultValue' in props && isValueValid(props.defaultValue)) {
    return [].concat(props.defaultValue);
  }

  if ('wheelDefaultValue' in props && isValueValid(props.wheelDefaultValue)) {
    return [].concat(props.wheelDefaultValue);
  }

  return defaultValue;
};

export const normalState = (props) => {
  const { valueMember, dataSource } = props;
  const value = getValues(
    props,
    dataSource!.map((item) => item[0] && item[0][valueMember!]),
  );
  return {
    value,
    objValue: props.dataSource.map(
      (item, index) => item.filter((d) => d[valueMember!] === value[index])[0],
    ),
    dataSource: props.dataSource,
  };
};

const cascaderState = (props) => {
  const { valueMember, cols } = props;
  let newValues = getValues(props, []);
  const newObjValues: any[] = [];
  const newDateSource: any[] = [];

  const parseLevel = ({ level = 0, dataSource }) => {
    newDateSource[level] = dataSource.map((item, index) => {
      const { children, ...others } = item;
      if (
        // eslint-disable-next-line operator-linebreak
        (isValueValid(newValues[level]) && item[valueMember!] === newValues[level]) ||
        (!isValueValid(newValues[level]) && index === 0)
      ) {
        newValues[level] = item[valueMember!];
        newObjValues[level] = others;

        if (isArray(children) && children.length > 0 && level + 1 < cols!) {
          parseLevel({
            level: level + 1,
            dataSource: children,
          });
        }
      }

      return others;
    });

    return newValues;
  };

  newValues = parseLevel({ dataSource: props.dataSource });

  return {
    value: newValues,
    objValue: newObjValues,
    dataSource: newDateSource,
  };
};

export default {
  getSource(props) {
    return isCascader(props) ? cascaderState(props) : normalState(props);
  },
  isValueValid,
};
