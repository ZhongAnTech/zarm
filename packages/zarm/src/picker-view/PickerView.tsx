import React, { Component } from 'react';
import classnames from 'classnames';
import isEqual from 'lodash/isEqual';
import Wheel from '../wheel';
import { isCascader } from '../utils/validate';
import parseProps from './utils/parseProps';
import removeFnFromProps from './utils/removeFnFromProps';
import type { BasePickerViewProps } from './PropsType';
import type { WheelValue, WheelItem } from '../wheel/interface';

export interface PickerViewProps extends BasePickerViewProps {
  prefixCls?: string;
  className?: string;
}

interface PickerViewState {
  value: Array<WheelValue>;
  dataSource: WheelItem[][];
}

export default class PickerView extends Component<PickerViewProps, PickerViewState> {
  static defaultProps = {
    prefixCls: 'za-picker-view',
    dataSource: [],
    cols: Infinity,
    valueMember: 'value',
    itemRender: (data) => data.label,
    disabled: false,
  };

  state: PickerViewState = parseProps.getSource(this.props);

  // TODO: is this method still be used?
  static getDerivedStateFromProps(props, state) {
    if (
      !isEqual(
        removeFnFromProps(props, ['onChange']),
        removeFnFromProps(state.prevProps, ['onChange']),
      )
    ) {
      return {
        prevProps: props,
        ...parseProps.getSource(props),
      };
    }

    return null;
  }

  onValueChange = (selected: WheelValue, level: number) => {
    const value = this.state.value.slice();
    const { dataSource, onChange, valueMember, cols } = this.props;

    if (isCascader({ dataSource })) {
      value.length = level + 1;
    }
    value[level] = selected;
    const newState = parseProps.getSource({ dataSource, value, valueMember, cols });
    this.setState(newState);
    if (typeof onChange === 'function') {
      onChange(newState.objValue, level);
    }
  };

  renderWheel = () => {
    const { valueMember, itemRender, disabled, stopScroll } = this.props;
    const { dataSource, value } = this.state;

    return dataSource.map((item, index) => (
      <Wheel
        key={+index}
        dataSource={item}
        value={value[index]}
        valueMember={valueMember}
        itemRender={itemRender}
        disabled={disabled}
        onChange={(selected: WheelValue) => this.onValueChange(selected, index)}
        stopScroll={stopScroll}
      />
    ));
  };

  render() {
    const { prefixCls, className } = this.props;
    const cls = classnames(prefixCls, className);
    return (
      <div className={cls}>
        <div className={`${prefixCls}__content`}>{this.renderWheel()}</div>
        <div className={`${prefixCls}__mask ${prefixCls}__mask--top`} />
        <div className={`${prefixCls}__mask ${prefixCls}__mask--bottom`} />
      </div>
    );
  }
}
