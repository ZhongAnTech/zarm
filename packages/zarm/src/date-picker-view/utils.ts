import dayjs from 'dayjs';
import isLeapYear from 'dayjs/plugin/isLeapYear';
import isoWeek from 'dayjs/plugin/isoWeek';
import isoWeeksInYear from 'dayjs/plugin/isoWeeksInYear';
import isFunction from 'lodash/isFunction';
import * as React from 'react';
import { ConfigContext } from '../config-provider';
import { PickerColumn, PickerColumnItem } from '../picker-view';
import { ColumnType, COLUMN_TYPE, DatePickerFilter, MERIDIEM, RenderLabel } from './interface';

dayjs.extend(isoWeek);
dayjs.extend(isoWeeksInYear);
dayjs.extend(isLeapYear);

export const generateMeridiemHour = (hour: number, hasMeridiem: boolean) => {
  if (!hasMeridiem) return hour;
  if (hour === 0) hour = 12;
  if (hour > 12) hour -= 12;
  return hour;
};

export const useRenderLabel = (renderLabel?: RenderLabel): RenderLabel => {
  const { locale: globalLocale } = React.useContext(ConfigContext);
  const locale = globalLocale?.DatePickerView;

  return React.useCallback(
    (type, value) => {
      if (isFunction(renderLabel)) {
        return renderLabel(type, value);
      }

      switch (type) {
        case COLUMN_TYPE.YEAR:
          return value + locale.year;
        case COLUMN_TYPE.MONTH:
          return value + locale.month;
        case COLUMN_TYPE.WEEK:
          return value + locale.week;
        case COLUMN_TYPE.WEEK_DAY:
          return locale.weeks[value - 1];
        case COLUMN_TYPE.DAY:
          return value + locale.day;
        case COLUMN_TYPE.HOUR:
          return value + locale.hour;
        case COLUMN_TYPE.MINUTE:
          return value + locale.minute;
        case COLUMN_TYPE.SECOND:
          return value + locale.second;
        case COLUMN_TYPE.MERIDIEM:
          return value === MERIDIEM.AM ? locale.am : locale.pm;
        default:
          return value.toString();
      }
    },
    [renderLabel],
  );
};

export const padZero = (value: string | number, length = 2) => {
  return String(value).padStart(length, '0');
};

export const generateDatePickerColumns = (
  selectedValue: Date,
  min: Date,
  max: Date,
  columnType: ColumnType[],
  renderLabel: RenderLabel,
  filter?: DatePickerFilter,
) => {
  const hasMeridiem = columnType.includes(COLUMN_TYPE.MERIDIEM);

  const date = dayjs(selectedValue);
  const minDate = dayjs(min);
  const maxDate = dayjs(max);

  const minYear = min.getFullYear();
  const minMonth = min.getMonth() + 1;
  const minDay = min.getDate();
  const minWeek = minDate.isoWeek();
  const minWeekday = minDate.isoWeekday();
  const minHour = min.getHours();
  const minMinute = min.getMinutes();
  const minSecond = min.getSeconds();

  const maxYear = max.getFullYear();
  const maxMonth = max.getMonth() + 1;
  const maxDay = max.getDate();
  const maxWeek = maxDate.isoWeek();
  const maxWeekday = maxDate.isoWeekday();
  const maxHour = max.getHours();
  const maxMinute = max.getMinutes();
  const maxSecond = max.getSeconds();

  const selectedYear = date.year();
  const selectedMonth = date.month() + 1;
  const selectedDay = date.date();
  const selectedWeek = date.isoWeek();
  const selectedHour = date.hour();
  const selectedMinute = date.minute();
  const selectedYearWeeks = dayjs(`${selectedYear}/01/01`).isoWeeksInYear();

  const isInMinYear = selectedYear === minYear;
  const isInMaxYear = selectedYear === maxYear;
  const isInMinMonth = isInMinYear && selectedMonth === minMonth;
  const isInMaxMonth = isInMaxYear && selectedMonth === maxMonth;
  const isInMinDay = isInMinMonth && selectedDay === minDay;
  const isInMaxDay = isInMaxMonth && selectedDay === maxDay;
  const isInMinWeek = isInMinYear && selectedWeek === minWeek;
  const isInMaxWeek = isInMaxYear && selectedWeek === maxWeek;
  const isInMinHour = isInMinDay && selectedHour === minHour;
  const isInMaxHour = isInMaxDay && selectedHour === maxHour;
  const isInMinMinute = isInMinHour && selectedMinute === minMinute;
  const isInMaxMinute = isInMaxHour && selectedMinute === maxMinute;

  const firstDayInSelectedMonth = dayjs(date).startOf('month');

  const generateColumns = <T extends PickerColumnItem>(
    from: number,
    to: number,
    type: ColumnType,
    iteratee: (index: number) => T,
  ) => {
    let options: T[] = [];
    for (let i = from; i <= to; i++) {
      options.push(iteratee(i));
    }

    if (isFunction(filter)) {
      options = options.filter((option) =>
        filter(type, {
          get value() {
            return option.value as number;
          },
          get date() {
            return selectedValue;
          },
        }),
      );
    }

    return options;
  };

  const generateMeridiemHourColumn = (hour: number) => {
    if (!hasMeridiem) return hour;
    if (hour > 12) hour -= 12;
    return hour;
  };

  const columns: PickerColumn = [];

  columnType?.forEach((type) => {
    switch (type) {
      case COLUMN_TYPE.YEAR:
        columns.push(
          generateColumns(minYear, maxYear, COLUMN_TYPE.YEAR, (index) => ({
            value: index,
            label: renderLabel(COLUMN_TYPE.YEAR, index),
          })),
        );
        break;
      case COLUMN_TYPE.MONTH:
        columns.push(
          generateColumns(
            isInMinYear ? minMonth : 1,
            isInMaxYear ? maxMonth : 12,
            COLUMN_TYPE.MONTH,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.MONTH, index),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.WEEK:
        columns.push(
          generateColumns(
            isInMinYear ? minWeek : 1,
            isInMaxYear ? maxWeek : selectedYearWeeks,
            COLUMN_TYPE.WEEK,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.WEEK, index),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.WEEK_DAY:
        columns.push(
          generateColumns(
            isInMinWeek ? minWeekday : 1,
            isInMaxWeek ? maxWeekday : 7,
            COLUMN_TYPE.WEEK_DAY,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.WEEK_DAY, index),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.DAY:
        columns.push(
          generateColumns(
            isInMinMonth ? minDay : 1,
            isInMaxMonth ? maxDay : firstDayInSelectedMonth.daysInMonth(),
            COLUMN_TYPE.DAY,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.DAY, index),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.MERIDIEM:
        columns.push(
          generateColumns(MERIDIEM.AM, MERIDIEM.PM, COLUMN_TYPE.MERIDIEM, (index) => ({
            value: index,
            label: renderLabel(COLUMN_TYPE.MERIDIEM, index),
          })),
        );
        break;
      case COLUMN_TYPE.HOUR:
        columns.push(
          generateColumns(
            generateMeridiemHourColumn(isInMinDay ? minHour : 0),
            generateMeridiemHourColumn(isInMaxDay ? maxHour : 23),
            COLUMN_TYPE.HOUR,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.HOUR, generateMeridiemHour(index, hasMeridiem)),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.MINUTE:
        columns.push(
          generateColumns(
            isInMinHour ? minMinute : 0,
            isInMaxHour ? maxMinute : 59,
            COLUMN_TYPE.MINUTE,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.MINUTE, index),
            }),
          ),
        );
        break;
      case COLUMN_TYPE.SECOND:
        columns.push(
          generateColumns(
            isInMinMinute ? minSecond : 0,
            isInMaxMinute ? maxSecond : 59,
            COLUMN_TYPE.SECOND,
            (index) => ({
              value: index,
              label: renderLabel(COLUMN_TYPE.SECOND, index),
            }),
          ),
        );
        break;
    }
  });

  return columns;
};

export const dateToNumberArray = (value: Date, columnType: ColumnType[]) => {
  const date = dayjs(value);
  const hasWeek = columnType?.includes(COLUMN_TYPE.WEEK);
  const hasMeridiem = columnType?.includes(COLUMN_TYPE.MERIDIEM);
  return columnType?.map((type) => {
    switch (type) {
      case COLUMN_TYPE.YEAR:
        return hasWeek ? date.isoWeekYear() : date.year();
      case COLUMN_TYPE.MONTH:
        return date.month() + 1;
      case COLUMN_TYPE.DAY:
        return date.date();
      case COLUMN_TYPE.WEEK:
        return date.isoWeek();
      case COLUMN_TYPE.WEEK_DAY:
        return date.isoWeekday();
      case COLUMN_TYPE.MERIDIEM:
        return date.hour() >= 12 ? MERIDIEM.PM : MERIDIEM.AM;
      case COLUMN_TYPE.HOUR:
        return generateMeridiemHour(date.hour(), hasMeridiem);
      case COLUMN_TYPE.MINUTE:
        return date.minute();
      case COLUMN_TYPE.SECOND:
        return date.second();
      default:
        return undefined;
    }
  });
};

export const numberArrayToDate = (
  currentValue: Date,
  changedPickerValue: number[],
  columnType: ColumnType[],
  level: number,
) => {
  const date = dayjs(currentValue);
  const types = Object.fromEntries(
    columnType.map((type, index) => [type, changedPickerValue[index]]),
  );
  const currentColumnType = columnType[level];

  const getMeridiemHour = (meridiem: number, hour: number) => {
    if (meridiem === MERIDIEM.AM && hour >= 12) {
      hour -= 12;
    }
    if (meridiem === MERIDIEM.PM && hour < 12) {
      hour += 12;
    }
    return hour;
  };

  const getHour = (hour: number) => {
    const meridiemIndex = columnType?.findIndex((i) => i === COLUMN_TYPE.MERIDIEM);
    const meridiemValue = changedPickerValue[meridiemIndex] as number;
    return getMeridiemHour(meridiemValue, hour);
  };

  const year = types.year ?? date.isoWeekYear() ?? 1900;
  const month = types.month ?? date.month() + 1 ?? 0;
  const week = types.week ?? 1;
  const weekDay = types['week-day'] ?? 1;
  const day = types.day ?? date.date() ?? 1;
  const hour = types.hour ?? 0;
  const minute = types.minute ?? 0;
  const second = types.second ?? 0;

  const selectedMonth = month - 1;
  const selectedDay = Math.min(day, dayjs(new Date(year, selectedMonth, 1)).daysInMonth());
  const selectedHour = getHour(hour);
  const selectedDate = dayjs().hour(selectedHour).minute(minute).second(second);

  if (columnType.includes(COLUMN_TYPE.WEEK)) {
    if (currentColumnType === COLUMN_TYPE.MONTH) {
      return selectedDate.year(year).month(selectedMonth).toDate();
    }
    if (currentColumnType === COLUMN_TYPE.DAY) {
      return selectedDate.year(year).isoWeek(week).date(selectedDay).toDate();
    }
    return selectedDate.year(year).isoWeek(week).isoWeekday(weekDay).toDate();
  }

  return selectedDate.year(year).month(selectedMonth).date(selectedDay).toDate();
};
