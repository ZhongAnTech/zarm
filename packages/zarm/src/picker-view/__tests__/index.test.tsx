import React from 'react';
import { render, mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import BScroll from 'better-scroll';
import { mocked } from 'ts-jest/utils';
import Wheel from '../../wheel';
import PickerView from '../index';
import parseProps from '../utils/parseProps';
import type { PickerDataSource } from '../PropsType';

// TODO: make mock package for better-scroll package
const bsScrollInstance = {
  on: jest.fn().mockImplementationOnce((event, handler) => {
    if (event === 'scrollEnd') {
      handler();
    }
  }),
  destroy: jest.fn(),
  getSelectedIndex: jest.fn(),
  refresh: jest.fn(),
  wheelTo: jest.fn(),
  disable: jest.fn(),
  stop: jest.fn(),
};

jest.mock('better-scroll', () => {
  return jest.fn(() => bsScrollInstance);
});

const mockedBScroll = mocked(BScroll);

describe('PickerView', () => {
  afterAll(() => {
    jest.resetAllMocks();
  });
  afterEach(() => {
    jest.restoreAllMocks();
  });
  it('should not render wheels if datasource is empty', () => {
    const wrapper = shallow(<PickerView />);
    expect(wrapper.find('.za-picker-view__content').children()).toHaveLength(0);
  });

  it('should render wheels if datasource is not empty', () => {
    const dataSource: PickerDataSource = [
      { label: 'a', value: 1 },
      { label: 'b', value: 2 },
    ];
    const wrapper = shallow(<PickerView dataSource={dataSource} />);
    expect(wrapper.state('dataSource')).toEqual([
      [
        { label: 'a', value: 1 },
        { label: 'b', value: 2 },
      ],
    ]);
    expect(wrapper.find('.za-picker-view__content').children()).toHaveLength(1);
    expect(wrapper.find(Wheel).props()).toEqual(
      expect.objectContaining({
        dataSource: [
          { label: 'a', value: 1 },
          { label: 'b', value: 2 },
        ],
        value: 1,
        valueMember: 'value',
        itemRender: PickerView.defaultProps.itemRender,
        disabled: false,
        onChange: expect.any(Function),
        stopScroll: false,
      }),
    );
  });

  it('should render with correct className', () => {
    const wrapper = shallow(<PickerView className="foo" />);
    expect(wrapper.at(0).prop('className')).toEqual('za-picker-view foo');
    expect(wrapper.find('.za-picker-view__content')).toHaveLength(1);
    expect(wrapper.find('.za-picker-view__mask')).toHaveLength(2);
    expect(wrapper.find('.za-picker-view__mask--top')).toHaveLength(1);
    expect(wrapper.find('.za-picker-view__mask--bottom')).toHaveLength(1);
  });

  it('should handle value change event and re-render wheels with new props', () => {
    const getSourceSpy = jest.spyOn(parseProps, 'getSource');
    bsScrollInstance.getSelectedIndex.mockReturnValueOnce(1);

    mockedBScroll.mockReturnValueOnce(bsScrollInstance as any);
    const dataSource: PickerDataSource = [
      { label: 'a', value: 1 },
      { label: 'b', value: 2 },
    ];
    const props = { onChange: jest.fn(), dataSource };
    const wrapper = mount(<PickerView {...props} />);
    expect(bsScrollInstance.on).toBeCalledWith('scrollEnd', expect.any(Function));

    expect(getSourceSpy).toBeCalledWith({
      dataSource: props.dataSource,
      value: [2],
      valueMember: 'value',
      cols: Infinity,
    });

    expect(props.onChange).toBeCalledWith([{ label: 'b', value: 2 }], 0);
    expect(wrapper.state('value')).toEqual([2]);
    expect(wrapper.state('objValue')).toEqual([{ label: 'b', value: 2 }]);
    expect(wrapper.state('dataSource')).toEqual([
      [
        { label: 'a', value: 1 },
        { label: 'b', value: 2 },
      ],
    ]);
    expect(wrapper.find(Wheel).props()).toEqual(
      expect.objectContaining({
        dataSource: [
          { label: 'a', value: 1 },
          { label: 'b', value: 2 },
        ],
        value: 2,
        valueMember: 'value',
        itemRender: PickerView.defaultProps.itemRender,
        disabled: false,
        onChange: expect.any(Function),
        stopScroll: false,
      }),
    );
  });

  it('PickerView render visible', () => {
    const wrapper = render(
      <PickerView
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
        defaultValue="1"
        value="1"
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('PickerView disabled', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <PickerView
        disabled
        onChange={onChange}
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
      />,
    );
    expect(onChange).not.toBeCalled();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
