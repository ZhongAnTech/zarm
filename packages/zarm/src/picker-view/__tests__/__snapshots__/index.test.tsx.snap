// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickerView PickerView disabled 1`] = `
<PickerView
  cols={Infinity}
  dataSource={
    Array [
      Object {
        "label": "选项一",
        "value": "1",
      },
      Object {
        "label": "选项二",
        "value": "2",
      },
    ]
  }
  disabled={true}
  itemRender={[Function]}
  onChange={[MockFunction]}
  prefixCls="za-picker-view"
  valueMember="value"
>
  <div
    className="za-picker-view"
  >
    <div
      className="za-picker-view__content"
    >
      <Wheel
        dataSource={
          Array [
            Object {
              "label": "选项一",
              "value": "1",
            },
            Object {
              "label": "选项二",
              "value": "2",
            },
          ]
        }
        disabled={true}
        itemRender={[Function]}
        key="0"
        onChange={[Function]}
        prefixCls="za-wheel"
        stopScroll={false}
        value="1"
        valueMember="value"
      >
        <div
          className="za-wheel"
        >
          <div
            className="za-wheel__content"
          >
            <div
              className="za-wheel__item za-wheel__item--selected za-wheel__item--disabled"
              key="0"
            >
              选项一
            </div>
            <div
              className="za-wheel__item za-wheel__item--disabled"
              key="1"
            >
              选项二
            </div>
          </div>
        </div>
      </Wheel>
    </div>
    <div
      className="za-picker-view__mask za-picker-view__mask--top"
    />
    <div
      className="za-picker-view__mask za-picker-view__mask--bottom"
    />
  </div>
</PickerView>
`;

exports[`PickerView PickerView render visible 1`] = `
<div
  class="za-picker-view"
>
  <div
    class="za-picker-view__content"
  >
    <div
      class="za-wheel"
    >
      <div
        class="za-wheel__content"
      >
        <div
          class="za-wheel__item za-wheel__item--selected"
        >
          选项一
        </div>
        <div
          class="za-wheel__item"
        >
          选项二
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-picker-view__mask za-picker-view__mask--top"
  />
  <div
    class="za-picker-view__mask za-picker-view__mask--bottom"
  />
</div>
`;
